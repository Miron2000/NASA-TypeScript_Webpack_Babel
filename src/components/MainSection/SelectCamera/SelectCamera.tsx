import React, { FC } from 'react';
import { ObjRoverCameras } from '../../../types';

interface PropsForSelectCamera {
  roverChosen: string;
  value?: string;
  onChangeSelect?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
}

const SelectCamera: FC<PropsForSelectCamera> = ({ value, onChangeSelect, roverChosen }) => {
  const roverCameras: ObjRoverCameras = {
    spirit: [
      {
        abbrev: 'FHAZ',
        name: 'Front Hazard Avoidance Camera',
      },
      {
        abbrev: 'RHAZ',
        name: 'Rear Hazard Avoidance Camera',
      },
      {
        abbrev: 'NAVCAM',
        name: 'Navigation Camera',
      },
      {
        abbrev: 'PANCAM',
        name: 'Panoramic Camera',
      },
    ],
    curiosity: [
      {
        abbrev: 'FHAZ',
        name: 'Front Hazard Avoidance Camera',
      },
      {
        abbrev: 'RHAZ',
        name: 'Rear Hazard Avoidance Camera',
      },
      {
        abbrev: 'MAST',
        name: 'Mast Camera',
      },
      {
        abbrev: 'CHEMCAM',
        name: 'Chemistry and Camera Complex',
      },
      {
        abbrev: 'MARDI',
        name: 'Mars Descent Imager',
      },
      {
        abbrev: 'NAVCAM',
        name: 'Navigation Camera',
      },
    ],
    opportunity: [
      {
        abbrev: 'FHAZ',
        name: 'Front Hazard Avoidance Camera',
      },
      {
        abbrev: 'RHAZ',
        name: 'Rear Hazard Avoidance Camera',
      },
      {
        abbrev: 'NAVCAM',
        name: 'Navigation Camera',
      },
      {
        abbrev: 'PANCAM',
        name: 'Panoramic Camera',
      },
    ],
  };

  return (
    <>
      <p className="title_mainSection">Choose a Camera:</p>
      <div className="select__wrapper">
        <select value={value} onChange={onChangeSelect} className="select" data-testid="selectCamera__MainSection">
          {roverCameras[roverChosen as keyof ObjRoverCameras].map((obj) => <option key={obj.abbrev} value={obj.abbrev}>{obj.name}</option>)}
        </select>
      </div>
    </>
  );
};

export default SelectCamera;
