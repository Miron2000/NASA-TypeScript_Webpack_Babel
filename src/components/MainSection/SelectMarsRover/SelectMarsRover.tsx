import React, { FC } from 'react';

interface PropsForSelectMarsRover {
  value?: string;
  onChangeSelect?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
}

const SelectMarsRover: FC<PropsForSelectMarsRover> = ({ value, onChangeSelect }) => (
  <>
    <p className="title_mainSection">Choose a Mars rover:</p>
    <div className="select__wrapper">
      <select value={value} onChange={onChangeSelect} className="select select__marsRover" data-testid="selectMarsRover__MainSection">
        <option value="curiosity">Curiosity</option>
        <option value="opportunity">Opportunity</option>
        <option value="spirit">Spirit</option>
      </select>
    </div>
  </>
);

export default SelectMarsRover;
