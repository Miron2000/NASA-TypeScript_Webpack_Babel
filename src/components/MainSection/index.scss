$font-family: 'Encode Sans SC', sans-serif;

.section {
  .section__selects {
    display: flex;
    justify-content: space-evenly;

    .select__wrapper {
      position: relative;

      .select {
        -moz-appearance: none;
        -webkit-appearance: none;
        appearance: none;
        outline: none;
        font-family: $font-family;
        width: 100%;
        background-color: #fff;
        border-radius: 6px;
        border: 1px solid #27458d;
        padding: 10px 35px 10px 15px;
        cursor: pointer;

        &:focus {
          outline: none;
          box-shadow: 0 0 0 1px black, 0 0 10px red;
        }
      }

      &::before {
        content: '';
        position: absolute;
        top: 50%;
        right: 15px;
        transform: translateY(-50%);
        display: block;
        pointer-events: none;
        width: 0;
        height: 0;
        border-style: solid;
        border-width: 10.4px 6px 0 6px;
        border-color: #27458d transparent transparent transparent;
      }
    }
  }

  .btn__section {
    display: block;
    margin: 40px auto;
    text-transform: uppercase;
    outline: 0;
    background: #26458d;
    width: 10%;
    border: 0;
    padding: 15px;
    color: #FFFFFF;
    font-family: $font-family;
    font-size: 14px;
    cursor: pointer;
    &:hover{
      background: #da4b1e;
    }
  }
}
