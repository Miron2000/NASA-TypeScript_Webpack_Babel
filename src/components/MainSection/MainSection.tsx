import React, { useState } from 'react';
import fetchPhoto from '../../fetchPhoto';
import SelectMarsRover from './SelectMarsRover/SelectMarsRover';
import SelectCamera from './SelectCamera/SelectCamera';
import Pagination from './Pagination/Pagination';
import { PhotosArray } from '../../types';

const Preloader = require('../../preloader/preloader.gif');

const MainSection = () => {
  const [roverChosen, setRoverChosen] = useState<string>('curiosity');
  const [cameraChosen, setCameraChosen] = useState<string>('');
  const [photos, setPhotos] = useState<PhotosArray[]>([]);
  const [error, setError] = useState<string>('');
  const [loading, setLoading] = useState<boolean>(false);
  const [currentPage, setCurrentPage] = useState<number>(1);

  const handleRoverSelection = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setRoverChosen(event.target.value);
  };

  const handleCameraSelection = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setCameraChosen(event.target.value);
  };

  const handleSubmit = () => {
    setLoading(true);
    fetchPhoto(roverChosen, cameraChosen)
      .then((res: { error: React.SetStateAction<string>; photos: React.SetStateAction<PhotosArray[]>; }) => {
        if (res?.error) {
          setError(res.error);
          console.error(error);
          return;
        }
        setPhotos(res.photos);
        setLoading(false);
      });
  };

  const photosPerPage = 1;
  const currentPhoto = photos[currentPage - 1];

  const paginate = (pageNumber: number) => {
    setCurrentPage(pageNumber);
  };

  return (
    <section className="section">
      <div className="section__selects">
        <div>
          <SelectMarsRover value={roverChosen} onChangeSelect={handleRoverSelection} />
        </div>

        <div>
          <SelectCamera
            value={cameraChosen}
            onChangeSelect={handleCameraSelection}
            roverChosen={roverChosen}
          />
        </div>
      </div>
      <button type="button" className="btn__section" onClick={handleSubmit}>GO!</button>

      {loading ? <img className="preloader" src={Preloader} alt="preloader" /> : null}

      <div>
        {photos && currentPhoto ? <img className="photo__marsRover" key={currentPhoto?.id} src={currentPhoto?.img_src} alt="marsRover" /> : <img className="default__img" src="https://www.buro247.ru/local/share/images/73445.jpg" alt="default" />}
      </div>
      <Pagination photosPerPage={photosPerPage} totalPhotos={photos.length} paginate={paginate} currentPage={currentPage} setCurrentPage={setCurrentPage} />
    </section>
  );
};

export default MainSection;
