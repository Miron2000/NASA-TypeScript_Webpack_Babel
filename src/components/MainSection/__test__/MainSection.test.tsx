import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import MainSection from '../MainSection';
import SelectCamera from '../SelectCamera/SelectCamera';
import SelectMarsRover from '../SelectMarsRover/SelectMarsRover';

test('renders without crashing', async () => {
  const component = render(<MainSection />);
  expect(component).toMatchSnapshot();
  component.unmount();
});

test('header renders with correct text', () => {
  const component = render(<MainSection />);
  const headerElMainSection = component.getByTestId('header__MainSection');

  expect(headerElMainSection.textContent).toBe('Mars Rover Photos');
});

test('select SelectMarsRover contains initial value text', () => {
  const { getByTestId } = render(<SelectMarsRover />);

  const selectEl = (getByTestId('selectMarsRover__MainSection') as HTMLInputElement).value;
  expect(selectEl).toBe('curiosity');
});

test('select SelectCamera contains initial value text', () => {
  const { getByTestId } = render(<SelectCamera />);

  const selectEl = (getByTestId('selectCamera__MainSection') as HTMLInputElement).value;
  expect(selectEl).toBe('');
});

test('button renders with GO!', () => {
  const component = render(<MainSection />);
  const btn = component.getByTestId('btn__MainSection');

  expect(btn.textContent).toBe('GO!');
});

test('change value of select works correctly', () => {
  const { getByTestId } = render(<SelectMarsRover />);
  const selectEl = getByTestId('selectMarsRover__MainSection');
  const selectElValue = (getByTestId('selectMarsRover__MainSection') as HTMLInputElement).value;
  fireEvent.change(selectEl, {
    target: {
      value: 'opportunity',
    },
  });

  expect(selectElValue).toBe('opportunity');
});
