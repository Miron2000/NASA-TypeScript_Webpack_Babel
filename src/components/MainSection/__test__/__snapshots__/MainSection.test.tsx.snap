// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="section"
      >
        <h3
          data-testid="header__MainSection"
        >
          Mars Rover Photos
        </h3>
        <div
          class="section__selects"
        >
          <div>
            <p
              class="title_mainSection"
            >
              Choose a Mars rover:
            </p>
            <div
              class="select__wrapper"
            >
              <select
                class="select select__marsRover"
                data-testid="selectMarsRover__MainSection"
              >
                <option
                  value="curiosity"
                >
                  Curiosity
                </option>
                <option
                  value="opportunity"
                >
                  Opportunity
                </option>
                <option
                  value="spirit"
                >
                  Spirit
                </option>
              </select>
            </div>
          </div>
          <div>
            <p
              class="title_mainSection"
            >
              Choose a Camera:
            </p>
            <div
              class="select__wrapper"
            >
              <select
                class="select select__camera"
                data-testid="selectCamera__MainSection"
              >
                <option
                  value="FHAZ"
                >
                  Front Hazard Avoidance Camera
                </option>
                <option
                  value="RHAZ"
                >
                  Rear Hazard Avoidance Camera
                </option>
                <option
                  value="MAST"
                >
                  Mast Camera
                </option>
                <option
                  value="CHEMCAM"
                >
                  Chemistry and Camera Complex
                </option>
                <option
                  value="MARDI"
                >
                  Mars Descent Imager
                </option>
                <option
                  value="NAVCAM"
                >
                  Navigation Camera
                </option>
              </select>
            </div>
          </div>
        </div>
        <button
          class="btn__section"
          data-testid="btn__MainSection"
          type="button"
        >
          GO!
        </button>
        <div
          class="result__block-img"
        />
        <div />
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="section"
    >
      <h3
        data-testid="header__MainSection"
      >
        Mars Rover Photos
      </h3>
      <div
        class="section__selects"
      >
        <div>
          <p
            class="title_mainSection"
          >
            Choose a Mars rover:
          </p>
          <div
            class="select__wrapper"
          >
            <select
              class="select select__marsRover"
              data-testid="selectMarsRover__MainSection"
            >
              <option
                value="curiosity"
              >
                Curiosity
              </option>
              <option
                value="opportunity"
              >
                Opportunity
              </option>
              <option
                value="spirit"
              >
                Spirit
              </option>
            </select>
          </div>
        </div>
        <div>
          <p
            class="title_mainSection"
          >
            Choose a Camera:
          </p>
          <div
            class="select__wrapper"
          >
            <select
              class="select select__camera"
              data-testid="selectCamera__MainSection"
            >
              <option
                value="FHAZ"
              >
                Front Hazard Avoidance Camera
              </option>
              <option
                value="RHAZ"
              >
                Rear Hazard Avoidance Camera
              </option>
              <option
                value="MAST"
              >
                Mast Camera
              </option>
              <option
                value="CHEMCAM"
              >
                Chemistry and Camera Complex
              </option>
              <option
                value="MARDI"
              >
                Mars Descent Imager
              </option>
              <option
                value="NAVCAM"
              >
                Navigation Camera
              </option>
            </select>
          </div>
        </div>
      </div>
      <button
        class="btn__section"
        data-testid="btn__MainSection"
        type="button"
      >
        GO!
      </button>
      <div
        class="result__block-img"
      />
      <div />
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
